### ISSUE TITLE
In task 2, which files are we allowed to modify?

### ISSUE THREAD

**Issue by eri24816** [hpcac/2022-APAC-HPC-AI#1]
Hi. I'm a participant in the competition.

I understand that in the task Communications_Performance_with_UCX, we have to clone the repo, tune parameters or configurations by modifying the files in the repo, then submit the modified version. Yet, there seems to be no specifications of which files (or parts of the files) are we are allowed to change. Could you provide a list of the modifiable files?

Thank you.

**Comment by pentschev** [comment]
Please take a look at the [Objectives and Rules](https://github.com/hpcac/2022-APAC-HPC-AI/tree/main/Communications_Performance_with_UCX#objectives-and-rules):

> To achieve the expected results mentioned above, the user may modify cuDF, UCX, UCX-Py, and any other open-source library on the stack, both their configuration and source code is allowed.
>
> For configuration changes, the user is expected to submit patches to the provided scripts (e.g., launch.sh and run-cluster.sh) that can be executed in the same way as the original code is provided. If necessary, any required instructions must be submitted as well.
>
> If any code changes are made, the user must submit any patches and scripts for an automated build of the stack. For example, the original conda environment can be used as base, and the user-submitted scripts will build and install software overwriting the original packages in the conda environment.
>
> The user is not allowed to upgrade nor downgrade any libraries used. However, the user is allowed to back- and forward-port modifications of any of the respective libraries, if that modification is judged beneficial for performance by the user.
>
> The user is not allowed to make any changes to the problem itself. Any changes preserving equality of the input data generation, the problem's output and final presentation of results is allowed. If the final submission modifies input or output data, as well as how results are presented as an output of the provided code, the submission will be disqualified.

You are required to use the same software stack (cuDF, UCX, UCX-Py) on the same versions that were originally provided. To exemplify:

- Suppose you find an operation in UCX-Py (or cuDF, or UCX) that does not perform optimally and you can improve UCX-Py (or cuDF, or UCX) to provide better performance, this is *ALLOWED*;
- Suppose you find that a newer (or older) version of UCX-Py (or cuDF, or UCX) had better performance after (or before) modification X and you want to apply (or revert) that change in the same version of UCX-Py (or cuDF, or UCX) that is required for the competition, this is *ALLOWED*;
- Suppose you find that by upgrading (or downgrading) the version of UCX-Py (or cuDF, or UCX) and have your final submission with the requirement for that different version of UCX-Py (or cuDF, or UCX), this is *NOT ALLOWED AND WILL BE DISQUALIFIED*;
- Suppose you entirely remove the UCX-Py (or cuDF, or UCX-Py) by some other software stack that performs better for this particular problem, this is *NOT ALLOWED AND WILL BE DISQUALIFIED*.

You are also allowed to change pretty much any of the files provided with the exception of the problem execution in the `.py` files in https://github.com/hpcac/2022-APAC-HPC-AI/tree/main/Communications_Performance_with_UCX . However, you are still allowed to make changes to those files if, and only if, the nature of the problem itself does not change. To exemplify:

- Suppose you find an extra synchronization step that can be removed for better performance while preserving results, this is *ALLOWED*;
- Suppose you can reorder operations in such a way that you can optimize data transfer, this is *ALLOWED*;
- Suppose you remove a core step of the code, like the [`.merge()` call](https://github.com/hpcac/2022-APAC-HPC-AI/blob/ba69cd16c55021a8b5ff919a670ba752b8097819/Communications_Performance_with_UCX/cudf-merge.py#L128) that will definitely speed things up (as nothing is being processed then) but the results are not correct anymore, this is *NOT ALLOWED AND WILL BE DISQUALIFIED*.

To summarize, as long as you preserve the operations that are executed, while also respecting other requirements described in the [problem statement](https://github.com/hpcac/2022-APAC-HPC-AI/tree/main/Communications_Performance_with_UCX), and can prove that given the problem input the output can be deterministically verified, while continuing to use the same software stack (cuDF, UCX, UCX-Py) in the same provided versions, you are allowed to make changes to any piece of code.

I hope this clarifies the requirements, and feel free to respond here if you have additional questions.
